File: src/lib/editor/statblock-presets.ts
// Statblock Presets (#179) — so a GM stops typing `Armor Class:` for the two hundredth
// time.
//
// **A preset is verbatim fence text**, stamped copy-on-insert with no id left behind.
// That is the whole of what legitimises storing presets *outside* the ledger: no note
// ever depends on a preset existing, so a vault opened on a fresh machine loses a
// convenience and never a word of content. Any future "should this be app data?"
// question is measured against that test.
//
// Presets are **app-wide** — one JSON file in the app data directory, shared by every
// campaign on the machine — while the pointer at *which* one a vault stamps is that
// vault's preference. The store holds only GM-authored shapes; the two shipped presets
// below are code constants that are never written into it.
//
// **Resolution is silent.** `/statblock` stamps the vault's default, `/statblock <arg>`
// stamps the `<arg>` preset on an exact, case-insensitive, whole-argument match, and
// an unset or unresolvable pointer yields a blank statblock with no toast — a toast
// fires mid-fight, where the GM cannot act on it. The diagnosis lives in Settings.
//
// **Blank is not a preset but the absence of one**, which is why every function here
// returns `null` rather than a "Blank" entry: there is no shape to store for it, and
// an entry would need a row in Settings that could not be renamed or deleted.
//
// Nothing here cleans a value on the way in. It is not merely undesirable but
// *impossible*: pools count in opposite directions, so a clean `HP: 3/12` is `12/12`
// to one GM and `0/12` to another, and Grimoire cannot know which.
import {
  blankStatblock,
  parseStatblockBody,
  type Statblock,
} from "$lib/editor/statblock-block";

/**
 * One saved shape. The fence is exactly what the block serialized to — delimiters and
 * all — because the thing being stored is the text a GM would see in Obsidian, not a
 * parsed model that would have to be kept in step with the format.
 */
export interface StatblockPreset {
  name: string;
  fence: string;
}

// ─── The two shipped shapes ───────────────────────────────────────────────────
//
// Read-only code constants, never injected into the store. They are what a GM meets
// before authoring anything of their own, so between them they have to cover both
// halves of what a preset is for: a system's labels, and a worked example.

/** The fence delimiters, written once so the constants below read as their bodies. */
const fence = (body: string) => ["```statblock", body, "```"].join("\n");

/**
 * A bare skeleton of the labels a 5E statblock carries, and no values at all. It
 * belongs to a system, so it says nothing Grimoire had to invent.
 */
const SRD_5E: StatblockPreset = {
  name: "5E SRD",
  fence: fence(
    [
      "Armor Class:",
      "Hit Points:",
      "Speed:",
      "STR:",
      "DEX:",
      "CON:",
      "INT:",
      "WIS:",
      "CHA:",
      "Saving Throws:",
      "Skills:",
      "Senses:",
      "Languages:",
      "Challenge:",
      "",
      "## Traits",
      "",
      "## Actions",
      "",
      "## Reactions",
    ].join("\n"),
  ),
};

/**
 * Grimoire's own teaching example, belonging to no system — its labels are ordinary
 * English words, not any game's vocabulary.
 *
 * It **carries values on purpose**. Playability is authored, not inferred (ADR-0016
 * §7), which leaves nothing on screen to teach a GM that a slash makes a pool — and
 * `HP: /` is just a slash. A GM who stamps this, clicks `12/12` and types `-5` has
 * learned the whole syntax without reading anything.
 *
 * **Four rows, four lessons**, and no more: a pool, a mark track, an inert value, and
 * a named prose entry. Deliberately not a playable creature — a real game outgrows it
 * in one session, and every extra row is a label a GM has to delete. No `Conditions:`
 * row, because a condition is not a Grimoire concept and a row named for one would
 * teach that it is.
 */
const LARGE_ORC: StatblockPreset = {
  name: "Large Orc",
  fence: fence(
    [
      "# Large Orc",
      "HP: 12/12",
      "Wounds: [ ][ ][ ]",
      "Defence: 14",
      "",
      "## Actions",
      "Greataxe: Hits hard enough to matter. Roll it however your game rolls.",
    ].join("\n"),
  ),
};

/** The presets that ship with Grimoire. Neither can be renamed, edited or deleted. */
export const SHIPPED_PRESETS: readonly StatblockPreset[] = [SRD_5E, LARGE_ORC];

// ─── Reading a preset ─────────────────────────────────────────────────────────

/**
 * A preset's fence as the body the format parses.
 *
 * Tolerant of a body with no delimiters, because the store is a JSON file a determined
 * GM may still hand-edit — fence text inside a JSON string makes that an unrealistic
 * route, not an impossible one, and a file that had been edited down to a bare body
 * should stamp rather than produce a statblock whose first row is ```` ```statblock ````.
 */
export function fenceBody(text: string): string {
  const lines = text.split("\n");
  if (lines[0]?.startsWith("```")) lines.shift();
  if (lines[lines.length - 1]?.trim() === "```") lines.pop();
  return lines.join("\n");
}

// ─── Choosing a preset ────────────────────────────────────────────────────────

/** The one name comparison this feature has: whole, trimmed, case-insensitive. */
export const sameName = (a: string, b: string) =>
  a.trim().toLowerCase() === b.trim().toLowerCase();

/** Whether a name belongs to a shipped preset, and so is not the GM's to take. */
export function isShippedName(name: string): boolean {
  return SHIPPED_PRESETS.some((preset) => sameName(preset.name, name));
}

/** Everything a GM can stamp: the shipped two, then their own. */
export function availablePresets(stored: StatblockPreset[]): StatblockPreset[] {
  return [...SHIPPED_PRESETS, ...stored];
}

/** The preset a name resolves to, or `null` — the whole of the matching rule. */
export function findPreset(
  presets: StatblockPreset[],
  name: string | null,
): StatblockPreset | null {
  const wanted = name?.trim();
  if (!wanted) return null;
  return presets.find((preset) => sameName(preset.name, wanted)) ?? null;
}

/**
 * The preset a `/statblock` stamps, or `null` for a blank one.
 *
 * The argument's match is **exact and whole**, not loose, and the failure modes are
 * why: a miss falls through to the vault's default — the shape the GM chose, one
 * restamp from fixed — while a loose misfire would silently hand over a shape nobody
 * asked for. Because the store is app-wide, a loose match would also let a preset
 * named `Orc` shadow `Orc Warlord` in every campaign on the machine.
 */
export function resolvePreset(
  argument: string,
  presets: StatblockPreset[],
  defaultName: string | null,
): StatblockPreset | null {
  return findPreset(presets, argument) ?? findPreset(presets, defaultName);
}

/**
 * The attributes `/statblock` inserts — a copy of the preset's shape and nothing else.
 * No id, no reference, nothing pointing back: editing or deleting a preset afterwards
 * cannot reach a note that was stamped from it.
 *
 * **The argument is always the title**, whether it matched a preset or not — never the
 * `# Name` inside the preset's own fence, which is what the preset's author called it
 * rather than what this GM just did.
 *
 * The one place the argument is not copied character for character is its *casing*,
 * and only when it matched: `/statblock goblin` titles the block `# Goblin`, because
 * the match was case-insensitive and the GM plainly meant that preset. A miss keeps
 * the argument exactly as typed — there is no name to prefer, and inventing a casing
 * for it would be Grimoire rewriting what the GM wrote.
 */
export function statblockFromPreset(
  preset: StatblockPreset | null,
  argument: string,
): Statblock {
  const block = preset
    ? parseStatblockBody(fenceBody(preset.fence))
    : blankStatblock();
  const name = argument.trim();
  if (!name) return block;
  const matched = preset && sameName(preset.name, name);
  return { ...block, name: matched ? preset.name.trim() : name };
}

File: src-tauri/src/commands/statblock_presets.rs
//! Statblock Presets (#179) — GM-authored statblock shapes, held app-wide.
//!
//! **Why these live outside the ledger.** A preset is verbatim fence text, stamped
//! copy-on-insert with no id left behind, so no note ever depends on one existing. A
//! vault opened on a fresh machine loses a convenience and never a word of content.
//! That test — and only that test — is what makes app data the right home; anything a
//! note could depend on belongs in the ledger.
//!
//! **One file rather than many**, against the usual recommendation, so it can be
//! emailed or dropped into a sync folder. The accepted cost is a one-way door: fence
//! text inside a JSON string makes hand-editing not a realistic route, and unlike a
//! raw fence in a note there is no Obsidian escape hatch for app data. That is what
//! makes complete in-app authoring load-bearing rather than merely nice.
//!
//! **Only GM-authored shapes are stored.** The two shipped presets are frontend code
//! constants and are never written here, which is what keeps them un-editable and
//! un-deletable without a flag on every row.
//!
//! Nothing here understands a fence. The text is opaque: no parsing, and above all no
//! value-cleaning — pools count in opposite directions, so a clean `HP: 3/12` is
//! `12/12` to one GM and `0/12` to another, and the backend cannot know which.
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Path, PathBuf};
use tauri::AppHandle;
use tauri::Manager;

const PRESETS_FILE: &str = "statblock-presets.json";

/// One saved shape: the name a GM stamps it by, and the fence text verbatim.
#[derive(Debug, Serialize, specta::Type, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct StatblockPreset {
    pub name: String,
    pub fence: String,
}

/// The file's shape. An object rather than a bare array so a later key can be added
/// without every existing file becoming unreadable.
#[derive(Debug, Serialize, Deserialize, Default, PartialEq)]
#[serde(rename_all = "camelCase", default)]
struct PresetStore {
    presets: Vec<StatblockPreset>,
}

/// Names are compared the way `/statblock <arg>` matches them: case-insensitively,
/// whole, and with surrounding whitespace ignored.
fn same_name(a: &str, b: &str) -> bool {
    a.trim().eq_ignore_ascii_case(b.trim())
}

fn presets_path(app: &AppHandle) -> Result<PathBuf, String> {
    let data_dir = app
        .path()
        .app_data_dir()
        .map_err(|e| format!("Failed to resolve app data dir: {}", e))?;
    fs::create_dir_all(&data_dir).map_err(|e| format!("Failed to create app data dir: {}", e))?;
    Ok(data_dir.join(PRESETS_FILE))
}

fn read_store(path: &Path) -> Result<PresetStore, String> {
    if !path.exists() {
        return Ok(PresetStore::default());
    }
    let contents =
        fs::read_to_string(path).map_err(|e| format!("Failed to read statblock presets: {}", e))?;
    serde_json::from_str(&contents).map_err(|e| format!("Failed to parse statblock presets: {}", e))
}

fn write_store(path: &Path, store: &PresetStore) -> Result<(), String> {
    let contents = serde_json::to_string_pretty(store)
        .map_err(|e| format!("Failed to serialize statblock presets: {}", e))?;
    fs::write(path, contents).map_err(|e| format!("Failed to write statblock presets: {}", e))
}

/// Save over an existing name, or append. The GM's order is otherwise untouched —
/// re-saving a shape must not move its row in Settings.
fn upsert(store: &mut PresetStore, preset: StatblockPreset) {
    match store
        .presets
        .iter_mut()
        .find(|p| same_name(&p.name, &preset.name))
    {
        Some(existing) => *existing = preset,
        None => store.presets.push(preset),
    }
}

fn rename(store: &mut PresetStore, from: &str, to: &str) -> Result<(), String> {
    if to.trim().is_empty() {
        return Err("ERR_EMPTY_NAME: a preset needs a name".into());
    }
    if store
        .presets
        .iter()
        .any(|p| same_name(&p.name, to) && !same_name(&p.name, from))
    {
        return Err("ERR_NAME_TAKEN: a preset of that name already exists".into());
    }
    match store.presets.iter_mut().find(|p| same_name(&p.name, from)) {
        Some(preset) => {
            preset.name = to.trim().to_string();
            Ok(())
        }
        None => Err(format!("No statblock preset named {}", from)),
    }
}

// ─── Commands ─────────────────────────────────────────────────────────────────

#[tauri::command]
#[specta::specta]
pub fn list_statblock_presets(app: AppHandle) -> Result<Vec<StatblockPreset>, String> {
    Ok(read_store(&presets_path(&app)?)?.presets)
}

/// Save a shape under a name, overwriting any shape already saved under it. The fence
/// is stored exactly as given — this is the whole of "save shape as preset".
#[tauri::command]
#[specta::specta]
pub fn save_statblock_preset(app: AppHandle, name: String, fence: String) -> Result<(), String> {
    if name.trim().is_empty() {
        return Err("ERR_EMPTY_NAME: a preset needs a name".into());
    }
    let path = presets_path(&app)?;
    let mut store = read_store(&path)?;
    upsert(
        &mut store,
        StatblockPreset {
            name: name.trim().to_string(),
            fence,
        },
    );
    write_store(&path, &store)
}

#[tauri::command]
#[specta::specta]
pub fn rename_statblock_preset(app: AppHandle, from: String, to: String) -> Result<(), String> {
    let path = presets_path(&app)?;
    let mut store = read_store(&path)?;
    rename(&mut store, &from, &to)?;
    write_store(&path, &store)
}

/// Delete a shape. No note can be affected: a stamped block holds a copy and no
/// reference, so this reaches nothing but the convenience itself.
#[tauri::command]
#[specta::specta]
pub fn delete_statblock_preset(app: AppHandle, name: String) -> Result<(), String> {
    let path = presets_path(&app)?;
    let mut store = read_store(&path)?;
    store.presets.retain(|p| !same_name(&p.name, &name));
    write_store(&path, &store)
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    fn preset(name: &str, fence: &str) -> StatblockPreset {
        StatblockPreset {
            name: name.into(),
            fence: fence.into(),
        }
    }

    #[test]
    fn roundtrip_preserves_fence_text_verbatim() {
        let dir = tempdir().unwrap();
        let path = dir.path().join(PRESETS_FILE);
        let fence = "```statblock\n# Goblin\nHP: 3/12\nWounds: [x][ ]\n```";

        let store = PresetStore {
            presets: vec![preset("Goblin", fence)],
        };
        write_store(&path, &store).unwrap();

        assert_eq!(read_store(&path).unwrap(), store);
        assert_eq!(read_store(&path).unwrap().presets[0].fence, fence);
    }

    #[test]
    fn missing_file_yields_an_empty_store() {
        let dir = tempdir().unwrap();
        assert_eq!(
            read_store(&dir.path().join(PRESETS_FILE)).unwrap(),
            PresetStore::default()
        );
    }

    #[test]
    fn upsert_overwrites_by_case_insensitive_name_and_keeps_position() {
        let mut store = PresetStore {
            presets: vec![preset("Goblin", "a"), preset("Orc", "b")],
        };
        upsert(&mut store, preset("goblin", "c"));

        assert_eq!(store.presets.len(), 2);
        assert_eq!(store.presets[0].name, "goblin");
        assert_eq!(store.presets[0].fence, "c");
        assert_eq!(store.presets[1].name, "Orc");
    }

    #[test]
    fn upsert_appends_a_new_name() {
        let mut store = PresetStore {
            presets: vec![preset("Goblin", "a")],
        };
        upsert(&mut store, preset("Orc", "b"));
        assert_eq!(
            store
                .presets
                .iter()
                .map(|p| p.name.as_str())
                .collect::<Vec<_>>(),
            vec!["Goblin", "Orc"]
        );
    }

    #[test]
    fn rename_refuses_a_taken_name() {
        let mut store = PresetStore {
            presets: vec![preset("Goblin", "a"), preset("Orc", "b")],
        };
        let err = rename(&mut store, "Goblin", "orc").unwrap_err();
        assert!(err.starts_with("ERR_NAME_TAKEN"));
        assert_eq!(store.presets[0].name, "Goblin");
    }

    #[test]
    fn rename_allows_a_change_of_case_on_the_same_preset() {
        let mut store = PresetStore {
            presets: vec![preset("goblin", "a")],
        };
        rename(&mut store, "goblin", "Goblin").unwrap();
        assert_eq!(store.presets[0].name, "Goblin");
    }

    #[test]
    fn rename_refuses_an_empty_name() {
        let mut store = PresetStore {
            presets: vec![preset("Goblin", "a")],
        };
        assert!(
            rename(&mut store, "Goblin", "   ")
                .unwrap_err()
                .starts_with("ERR_EMPTY_NAME")
        );
    }
}
